helix-view/src/document.rs

```diff
@@ -150,13 +156,63 @@ pub struct Document {
     diff_handle: Option<DiffHandle>,
 }
 
+/// Inlay hints for a single `(Document, View)` combo.
+///
+/// There are `*_inlay_hints` field for each kind of hints an LSP can send since we offer the
+/// option to style theme differently in the theme according to the (currently supported) kinds
+/// (`type`, `parameter` and the rest).
+///
+/// Inlay hints are always `InlineAnnotation`s, not overlays or line-ones: LSP may choose to place
+/// them anywhere in the text and will sometime offer config options to move them where the user
+/// wants them but it shouldn't be Helix who decides that so we use the most precise positioning.
+///
+/// The padding for inlay hints is stored directly before/after the hint text itself. This is
+/// necessary to ensure it's placed correctly instead of after all the time (or ignored because it
+/// came in a later layer and at the same `char_idx`).
+#[derive(Debug, Clone)]
+pub struct DocumentInlayHints {
+    /// Identifier for the inlay hints stored in this structure. To be checked to know if they have
+    /// to be recomputed on idle or not.
+    pub id: DocumentInlayHintsId,
+
+    /// Inlay hints of `TYPE` kind, if any.
+    pub type_inlay_hints: Rc<[InlineAnnotation]>,
+
+    /// Inlay hints of `PARAMETER` kind, if any.
+    pub parameter_inlay_hints: Rc<[InlineAnnotation]>,
+
+    /// Inlay hints that are neither `TYPE` nor `PARAMETER`.
+    ///
+    /// LSPs are not required to associate a kind to their inlay hints, for example Rust-Analyzer
+    /// currently never does (February 2023) and the LSP spec may add new kinds in the future that
+    /// we want to display even if we don't have some special highlighting for them.
+    pub other_inlay_hints: Rc<[InlineAnnotation]>,
+}
+
+/// Associated with a [`Document`] and [`ViewId`], uniquely identifies the state of inlay hints for
+/// for that document and view: if this changed since the last save, the inlay hints for the view
+/// should be recomputed.
+///
+/// We can't store the `ViewOffset` instead of the first and last asked-for lines because if
+/// softwrapping changes, the `ViewOffset` may not change while the displayed lines will.
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct DocumentInlayHintsId {
+    /// Document revision at which the inlay hints were requested.
+    pub revision: usize,
```

Comment:
I tested out this PR some more and noticed that using the revision for refreshing doesn't work as well as I had hoped (it cases issues with undo for example). It seemed like a good solution at first but I cameup with something better/easier: Since all inline annotations for all views need to be requested on idle timeout when the document is modified simply store a bool on the document:
``` rust
inlayhint_outdated: bool
```

* In `apply_impl` simply set this to ` true`.
* In `compute_inlay_hints_for_view`: reset this field to `false` and rerequest the inlay hints if this is true OR the view position changed